scripts/bluetooth/mesh/mesh_dfu_metadata.py

```
@@ -0,0 +1,388 @@
+FILE_NAME = 'dfu_application.zip_ble_mesh_metadata.json'
+
+
+def exit_with_error_msg(msg):
+    print("Extracting BLE Mesh metadata failed with reason:")
+    print(f"\t{msg}")
+    print("You can bypass this script by disabling the CONFIG_BT_MESH_DFU_METADATA_ON_BUILD option in you project config")
+    sys.exit(0)
+
+
+class Elem:
+    def __init__(self, loc):
+        self.loc = loc
+        self.vnd_list = []
+        self.sig_list = []
+
+    def vnd_model_add(self, cid, vid):
+        self.vnd_list.append((cid << 16) + vid)
+
+    def sig_model_add(self, id):
+        self.sig_list.append(id)
+
+    def bytestring_generate(self):
+        bytestring = bytearray()
+        bytestring.extend(self.loc.to_bytes(2, 'little'))
+        bytestring.append(len(self.sig_list))
+        bytestring.append(len(self.vnd_list))
+
+        for sig in self.sig_list:
+            bytestring.extend(sig.to_bytes(2, 'little'))
+        for vnd in self.vnd_list:
+            bytestring.extend(vnd.to_bytes(4, 'little'))
+
+        return bytestring
+
+    def dict_generate(self):
+        return {
+            "location": self.loc,
+            "sig_models": self.sig_list,
+            "vendor_models": self.vnd_list,
+        }
+
+
+class Comp0:
+    # Must stay in order
+    FEATURE_KCONF_OPTS = [
+        'CONFIG_BT_MESH_RELAY',
+        'CONFIG_BT_MESH_GATT_PROXY',
+        'CONFIG_BT_MESH_FRIEND',
+        'CONFIG_BT_MESH_LOW_POWER',
+    ]
+
+    def __init__(self, cid, pid, vid, kconfig):
+        self.elems = []
+        self.cid = cid
+        self.pid = pid
+        self.vid = vid
+        self.__features_add(kconfig)
+        self.hash = None
+
+    def elem_add(self, loc):
+        new_elem = Elem(loc)
+        self.elems.append(new_elem)
+        return new_elem
+
+    def __features_add(self, kconfig):
+        if 'CONFIG_BT_MESH_CRPL' not in kconfig.keys():
+            exit_with_error_msg("Could not find CONFIG_BT_MESH_CRPL Kconfig option")
```

Could it be better to raise the exception up to the main function and catch it there?

---

? :(

---

Sorry, got a bit confused. It should be better now.